src/datadog/trace_segment.cpp

```diff
@@ -389,37 +414,32 @@ void TraceSegment::write_sampling_delegation_response(DictWriter& writer) {
     j["mechanism"] = *sampling_decision_->mechanism;
   }
 
-  if (expecting_delegated_sampling_decision_) {
-    expecting_delegated_sampling_decision_ = false;
-    trace_tags_.emplace_back(tags::internal::sampling_decider, "1");
-  }
-
-  writer.set(DD_SAMPLING_DELEGATION_RESPONSE_HEADER, j.dump());
-  return;
+  writer.set(sampling_delegation_response_header, j.dump());
+  sampling_delegation_.sent_response_header = true;
 }
 
 Expected<void> TraceSegment::read_sampling_delegation_response(
     const DictReader& headers) {
-  if (auto sampling_delegation_response =
-          headers.lookup(DD_SAMPLING_DELEGATION_RESPONSE_HEADER)) {
-    auto sampling_decision =
-        parse_sampling_delegation_response(*sampling_delegation_response);
-    if (!sampling_decision) {
-      return sampling_decision.error();
-    }
+  auto header_value = headers.lookup(sampling_delegation_response_header);
+  if (!header_value) {
+    return {};
+  }
 
-    {
-      std::lock_guard<std::mutex> lock(mutex_);
-      sampling_decision_ = *sampling_decision;
-      expecting_delegated_sampling_decision_ = false;
+  auto decision = parse_sampling_delegation_response(*header_value);
+  if (Error* error = decision.if_error()) {
+    return std::move(*error);
+  }
 
-      // Successfully delegated the sampling decision, set the sampling decider
-      // tag only if it is the root span, to distinguish the root from the
-      // decider
-      if (spans_.front()->parent_id == 0) {
-        trace_tags_.emplace_back(tags::internal::sampling_decider, "0");
-      }
-    }
+  sampling_delegation_.received_matching_response_header = true;
+
+  std::lock_guard<std::mutex> lock(mutex_);
+  // Overwrite any existing sampling decision if and only if the existing
+  // decision is not a local manual override.
+  if (!(sampling_decision_ &&
+        sampling_decision_->origin == SamplingDecision::Origin::LOCAL &&
+        sampling_decision_->mechanism == int(SamplingMechanism::MANUAL))) {
```

Comment:
Feel free to put the `static_cast` back (instead of my function-style C cast).

Reply:
Nah, not for numeric and enum conversions. We can revisit this across the entire repo another time. For now I'll keep in consistent with other uses.

Compare the output of the following two commands:
```
$ git grep '\<int('
``` 
```
$ git grep static_cast src/ ':!*json.hpp'
```